libs/icu_headroom_metric.py

```diff
@@ -0,0 +1,168 @@
+
+
+class CovidPatientsMethod(enum.Enum):
+    ACTUAL = 0
+    ESTIMATED = 1
+
+
+class NonCovidPatientsMethod(enum.Enum):
+    ACTUAL = 0
+    ESTIMATED_FROM_TYPICAL_UTILIZATION = 1
+    ESTIMATED_FROM_TOTAL_ICU_ACTUAL = 2
+
+
+DEFAULT_ICU_DECOMP = 0.21
+
+ICU_DECOMP_OVERRIDE = {
+    "AL": 0.15,
+    "AZ": 0.4,
+    "DE": 0.3,
+    "DC": 0.15,
+    "GA": 0.1,
+    # TODO(https://trello.com/c/1ddB5ntl/): CCM is currently giving us an
+    # extra-high utilization rate. If that gets fixed we may need to bump this
+    # back down.
+    "MS": 0.37,
+    "NV": 0.25,
+    "RI": 0,
+}
+
+
+def get_decomp_for_state(state: str) -> float:
+    return ICU_DECOMP_OVERRIDE.get(state, DEFAULT_ICU_DECOMP)
+
+
+class ICUMetricData:
+    def __init__(self, data, latest_values, decomp_factor):
+        self.data = data
+        self.latest_values = latest_values
+        self.decomp_factor = decomp_factor
+
+    @property
+    def actual_current_icu_covid(self) -> Optional[pd.Series]:
+        actuals = self.data[CommonFields.CURRENT_ICU]
+        if not actuals.any():
+            return None
+
+        return actuals
+
+    @property
+    def estimated_current_icu_covid(self) -> Optional[pd.Series]:
+        estimated = self.data[schema.CURRENT_ICU]
+        if not estimated.any():
+            return None
+
+        return estimated
+
+    @property
+    def actual_current_icu_total(self) -> Optional[pd.Series]:
+        actuals = self.data[CommonFields.CURRENT_ICU_TOTAL]
+        if not actuals.any():
+            return None
+
+        return actuals
+
+    @property
+    def total_icu_beds(self):
+        timeseries = self.data[CommonFields.ICU_BEDS]
+        if timeseries.any():
+            return timeseries
+        return self.latest_values[CommonFields.ICU_BEDS]
+
+    @property
+    def typical_usage_rate(self):
+        return self.latest_values[CommonFields.ICU_TYPICAL_OCCUPANCY_RATE]
+
+    @property
+    def current_icu_non_covid_with_source(self):
+        if self.actual_current_icu_covid is not None and self.actual_current_icu_total is not None:
+            source = NonCovidPatientsMethod.ACTUAL
+            return (self.actual_current_icu_total - self.actual_current_icu_covid, source)
+
+        if self.actual_current_icu_total is not None:
+            source = NonCovidPatientsMethod.ESTIMATED_FROM_TOTAL_ICU_ACTUAL
+            return (self.actual_current_icu_total - self.estimated_current_icu_covid, source)
+
+        # Do decomp
+        non_covid_utilization = self.typical_usage_rate - self.decomp_factor
+        source = NonCovidPatientsMethod.ESTIMATED_FROM_TYPICAL_UTILIZATION
+
+        if isinstance(self.total_icu_beds, pd.Series):
+            decomp = non_covid_utilization * self.total_icu_beds
+            decomp.name = None
+            return (decomp, source)
+
+        total_icu_beds = pd.Series([self.total_icu_beds] * len(self.data), index=self.data.index)
+        decomp = non_covid_utilization * total_icu_beds
```

Comment:
1. Would this code be simpler if you changed self.total_icu_beds to always be a series?  (just extend the scalar into a series if we only have the latest value)
2. "decomp" seems like a not-super-useful variable name.  Seems like it should be `non_covid_patients` or something.  I guess it could be `non_covid_patients_via_decomp` but seems unnecessary?